### ISSUE TITLE
Document how to create custom external provider authentication

### ISSUE THREAD

**Issue by travbeamo** [dotnet/aspnetcore#25029]
How can a custom external authentication provider (for example called **MyCustomAuth**) be implemented (from start to finish) like:
* [Your existing providers (Facebook/Google/Microsoft/etc..)](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/social/?view=aspnetcore-3.1&tabs=visual-studio) 
* [3rd-party providers (GitHub/LinkedIn/Yahoo/etc..)](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/social/other-logins?view=aspnetcore-3.1)

I am able to setup the WebApp side & get an external login button, but cannot get the external provider side working with how the endpoints (**authorize**, **token** & **user information**) work and interact back to the WebApp.

I have attempted using an OAuthHandler, but if this is overkill and and a RemoteAuthenticationHandler or AuthenticationHandler could be used instead, please let me know.

Is there a basic example project to show how both sides work together OR in one that particularly demonstrates how the endpoints work with the authentication middleware and gets from:
* page **/Identity/Account/Login**, where a user then clicks on **MyCustomAuth** button under "**Use another service to log in**"
![image](https://user-images.githubusercontent.com/63974155/90575816-9de28680-e1ff-11ea-92c2-197f0879af38.png)
* to page **/Identity/Account/ExternalLogin**, where user enters their email and clicks **Register** under "**Associate your account**".
![image](https://user-images.githubusercontent.com/63974155/90575864-c10d3600-e1ff-11ea-910b-9d9ed21f3c6a.png)

I think this documentation & example project would be very helpful. 

Thanks in advance.

**Comment by Tratcher** [comment]
We could certainly cover some high level scenarios and and common protocols like OAuth, but the details depend a lot on the kind of authentation you're trying to implement.

Can you summarize the expected flow for said custom auth?

**Comment by tb-mtg** [comment]
Expectations are:
1. On page **/Identity/Account/Login**, user is redirected to another external login page after on clicking **MyCustomAuth** button
2. User accept that it is with an external provider, then submits form with username & password
3. Externally validate username & password (which I have working method for doing). 
a. Failed attempts are returned back to external login.
b. Successful attempts returns a id/key/code for that is unique for that user in the external login provider.  This id/key/code is to be used as the **ProviderKey** value in the AspNetUserLogins table 

--This is where I'm stuck--

4. Somehow user is returned to **/Identity/Account/ExternalLogin** on WebApp and process continues as it would  when  existing implementations (Facebook,Goggle, etc.) where the **ProviderKey** has been added to the AspNetUserLogins table.
a. New users must submit form with email to complete creating a local WebApp account.
5. Existing users are now signed-in and redirected to the ReturnUrl on WebApp.

**That is the minimal basic expectation.** 

Ideally additional claims would also be returned from external provider as part of the process.  
For example an email address could be returned removing the need to prompt user for one in order to register. However the main priority for now is just a working custom authentication flow using AspNetCore Identity.

**Comment by Tratcher** [comment]
That flow as you describe it is indistinguishable from [OAuth2](https://github.com/dotnet/aspnetcore/blob/master/src/Security/Authentication/OAuth/src/OAuthHandler.cs).

Step 4 triggers here:
https://github.com/dotnet/aspnetcore/blob/55ce868cf9327e5b0d66bcc4ce31addeaed7ab6d/src/Security/Authentication/Core/src/RemoteAuthenticationHandler.cs#L140-L148
The user is stored in a temp cookie and the app is redirected back to the original challenge url (`/Identity/Account/ExternalLogin` in this example).

**Comment by travbeamo** [comment]
@Tratcher  Thanks for your help, much appreciated. However as that’s where I’m stuck, could you please provide a basic example of the endpoint(s) required to get to that trigger point you referred to where the temp cookie is created. 

The validation of username/password part in example can be trivial, I just need to see a working flow example so I can see where I’m going wrong.

**Comment by Tratcher** [comment]
Ok, I'm a little confused. You said you were sending them to an external login page, but then you're validating the username and password directly in the app? That's not auth middleware at all, that's just a separate page/action on your login controller that looks a lot like the existing one for Identity that accepts a username and password.
https://github.com/dotnet/aspnetcore/blob/36d887f250e4caad2dced1609ee2153d2100da48/src/Identity/UI/src/Areas/Identity/Pages/V4/Account/Login.cshtml.cs#L122-L132
